### PR TITLE
implement LowerExp and UpperExp for integers

### PR DESCRIPTION
**Pull request by maxbla**
Addresses https://github.com/rust-lang/rust/issues/39479

This implementation is heavily based on the preexisting `macro_rules! impl_Display` in the same file. I don't like the liberal use of unsafe in that macro and would like to modify it so `unsafe` is only present where necessary. What is Rust's policy on doing such modifications?

Also, I couldn't figure out where to put tests, can I have some help with that?

### CODE REVIEW

**Comment by Alexendoo** [comment]
Ping from triage, any updates? @kennytm

**Comment by kennytm** [comment]
@Alexendoo This is a Draft PR, it is waiting for author to click "Ready for review".

**Comment by maxbla** [comment]
Sorry, this is my first PR. I was waiting on an answer to the questions higher in the thread, as I knew it would not pass review without some changes. I have now marked it ready for review, because review will probably involve answers to the question of where to put tests.

**Comment by Dylan-DPC** [comment]
@kennytm if you can answer the above question(s) we can proceed with this PR :) thanks

**Comment by kennytm** [comment]
@maxbla Unit tests for libcore should be placed in `src/libcore/tests/`. For this PR, try [`src/libcore/tests/fmt/num.rs`](https://github.com/rust-lang/rust/blob/master/src/libcore/tests/fmt/num.rs).

File: src/libcore/fmt/num.rs
Comment (kennytm) [inline src/libcore/fmt/num.rs]: `impl` is insta-stable, i.e. the `#[unstable]` here has no effect and you can use `format!("{:e}", 123)` without `#![feature(IntegerExpFormat)]` anywhere (BTW feature names should be written like `integer_exp_format`). So this PR will additionally require FCP from @rust-lang/libs before merging.

Comment (maxbla) [reply src/libcore/fmt/num.rs]: What should I put here instead?

Comment (kennytm) [reply src/libcore/fmt/num.rs]: @maxbla `#[stable(feature = "integer_exp_format", since = "1.42.0")]`

File: src/libcore/tests/fmt/num.rs
Comment (kennytm) [inline src/libcore/tests/fmt/num.rs]: please add test cases involving:

* i32::MIN / i64::MIN / i128::MIN

Can (should) the new code handle precision flags?

```rust
assert_eq!(format!("{:.3e}", 123456789), "1.235e9");
```

Comment (maxbla) [reply src/libcore/tests/fmt/num.rs]: > Can (should) the new code handle precision flags?

It can 't as of 6b12eb9, but it definitely should. I completely overlooked this, thinking I could rely on `pad_integral` to deal with formatting flags.

**Comment by maxbla** [comment]
81305e2 adds precision support using `flt2dec::Part` and `fmt::pad_formatted_parts()`.

It feels a tad weird to use a module called `flt2dec` for integer formatting. I think `core::num::flt2dec::Part` might fit better in `core::fmt` and `core::num::flt2dec::Formatted` definitely would fit better there.

**Comment by maxbla** [comment]
`core::fmt::{pad_formatted_parts. write_formatted_parts}` both call `unsafe fn str::from_utf8_unchecked()` on their arguments, but are not marked as unsafe themselves. Is it possible this was done for a reason, or can I change that?

File: src/libcore/fmt/num.rs
Comment (KodrAus) [inline src/libcore/fmt/num.rs]: Before digging too deeply into the code itself can you offer any background on this implementation? Whether it's based on any common approaches used elsewhere?

**Comment by KodrAus** [comment]
> It feels a tad weird to use a module called flt2dec for integer formatting. I think core::num::flt2dec::Part might fit better in core::fmt and core::num::flt2dec::Formatted definitely would fit better there.

I think the original implementation came from an external library so it all ended up living together. I wouldn't be against refactoring out commonly useful pieces from it.

**Review by KodrAus — COMMENTED** [review]
Thanks for the PR @maxbla!

I've just left a few comments to start.

File: src/libcore/fmt/num.rs
Comment (maxbla) [reply src/libcore/fmt/num.rs]: Yeah, I'll inline it. It doesn't seem very reusable as-is. I thought breaking it out might improve readability.

Comment (maxbla) [reply src/libcore/fmt/num.rs]: It is heavily based on `impl_Display!` from the same file. It also makes use of some stuff from float's implementation (for precision).

File: src/libcore/fmt/num.rs
Comment (maxbla) [reply src/libcore/fmt/num.rs]: Yeah, I'll inline it. It doesn't seem very reusable as-is. I thought breaking it out might improve readability.

File: src/libcore/fmt/num.rs
Comment (maxbla) [reply src/libcore/fmt/num.rs]: done!

File: src/libcore/fmt/num.rs
Comment (KodrAus) [inline src/libcore/fmt/num.rs]: Let's order the update to `n` and `exponent` and the writing to the buffer and updating `curr` the same way in each of these blocks.

File: src/libcore/fmt/num.rs
Comment (KodrAus) [inline src/libcore/fmt/num.rs]: Ok, I can see what you were going for in the original implementation splitting out the initial exponent. I think it would be good to put this first part of the algorithm that does the initial work of figuring out precision in a block:

```rust
let (mut n, mut exponent, trailing_zeros, added_precision) = {
    ..
};
```

**Review by KodrAus — APPROVED** [review]
Alrighty, this is looking good to me! Thanks for working on this @maxbla!

Would you like to squash your commits down and I'll merge this one in?

**Comment by maxbla** [comment]
@KodrAus Thanks for the review. I'm excited to get my first contribution merged!
Is now a good time to move `core::num::flt2dec::Part` and `core::num::flt2dec::Formatted` to `core::fmt` or did I miss the window on that?

**Comment by maxbla** [comment]
I think this PR is now ready to merge.

**Comment by Dylan-DPC** [comment]
Not yet xD

We don't do merges here so it will be nice if you could revert the merge and rebase instead.

**Comment by maxbla** [comment]
Alright I think I did it right this time

**Comment by KodrAus** [comment]
Thanks for your patience @maxbla! This looks good to me.

@bors r+

**Pull request merged**
